cmd/drone-agent/agent.go

```diff
@@ -68,13 +69,34 @@ func loop(c *cli.Context) error {
 
 	// grpc.Dial(target, ))
 
+	agentKeepalive := keepalive.ClientParameters{}
+	if c.String("keepalive-time") != "" {
+		d, err := time.ParseDuration(c.String("keepalive-time"))
+
+		if err != nil {
+			return err
+		}
+
+		agentKeepalive.Time = d
+	}
+	if c.String("keepalive-timeout") != "" {
+		d, err := time.ParseDuration(c.String("keepalive-timeout"))
+
+		if err != nil {
+			return err
+		}
+
+		agentKeepalive.Timeout = d
+	}
+
 	conn, err := grpc.Dial(
 		c.String("server"),
 		grpc.WithInsecure(),
 		grpc.WithPerRPCCredentials(&credentials{
 			username: c.String("username"),
 			password: c.String("password"),
 		}),
+		grpc.WithKeepaliveParams(agentKeepalive),
```

Comment:
you should be able to simplify like this:

```
grpc.WithKeepaliveParams(
  keepalive.ClientParameters{
    Timeout: c.Duration("keepalive-timeout"),
    Time: c.Duration("keepalive-time"),
  },
),
```

If `c.Duration` is empty it will return the zero value, so this snippet here:

```
agentKeepalive := keepalive.ClientParameters{}
if dur := c.Duration("keepalive-time"); dur != 0 {
  agentKeepalive.Time = dur
}
```

should behave the same as this snippet here:

```
agentKeepalive := keepalive.ClientParameters{
  Time: c.Duration("keepalive-time"),
}
```

Reply:
So my only concern is I'm trying to maintain default values for `Time` (infinity) and `Timeout` (20s).

```
grpc.WithKeepaliveParams(
  keepalive.ClientParameters{
    Timeout: c.Duration("keepalive-timeout"),
    Time: c.Duration("keepalive-time"),
  },
),
```

would return `0` for both as defaults instead if user doesn't specify keep alive config.  Let me know if I'm wrong and I can implement it that way.  Otherwise I think I'll change what I have to:

```
agentKeepalive := keepalive.ClientParameters{}
if dur := c.Duration("keepalive-time"); dur != 0 {
  agentKeepalive.Time = dur
}
```

Reply:
I think this will still work as expected. There is no `nil` value for `time.Duration` so the grpc code will likely use infinity if the value is 0. I will look through the code to verify, though

Reply:
ok, looks like if `c.Duration` returns a zero value, the kubernetes defaults will kick in

```
	if kp.Time == 0 {
		kp.Time = defaultClientKeepaliveTime
	}
	if kp.Timeout == 0 {
		kp.Timeout = defaultClientKeepaliveTimeout
	}
```